sos/plugins/juju.py

```diff
@@ -76,37 +86,40 @@ def export_mongodb(self):
                 suggest_filename="{}.json".format(collection))
 
     def setup(self):
-        self.add_copy_spec([
-            "/var/lib/juju"
-        ])
         limit = self.get_option("log_size")
         self.add_copy_spec_limit("/var/log/upstart/juju-db.log",
                                  sizelimit=limit)
         self.add_copy_spec_limit("/var/log/upstart/juju-db.log.1",
                                  sizelimit=limit)
         if not self.get_option("all_logs"):
             # Capture the last bit of all files
-            for filename in os.listdir("/var/log/juju/"):
-                if filename.endswith(".log"):
-                    fullname = "/var/log/juju/" + filename
-                    self.add_copy_spec_limit(fullname, sizelimit=limit)
-            # Do just the all-machines from juju local
-            self.add_copy_spec_limit("/var/log/juju-*/all-machines.log",
-                                     sizelimit=limit)
+            if(os.path.isdir("/var/log/juju/")):
+                for filename in os.listdir("/var/log/juju/"):
+                    if filename.endswith(".log"):
+                        fullname = "/var/log/juju/" + filename
+                        self.add_copy_spec_limit(fullname, sizelimit=limit)
             self.add_cmd_output('ls -alRh /var/log/juju*')
+            self.add_cmd_output('ls -alRh /var/lib/juju/*')
+
         else:
             self.add_copy_spec([
                 "/var/log/juju",
-                "/var/log/juju-*"
+                "/var/log/juju-*",
+                "/var/lib/juju"
             ])
 
-        self.add_cmd_output([
-            "juju -v status",
-            "juju -v get-constraints"
-        ])
-
-        for service in self.get_deployed_services():
-            self.add_cmd_output("juju get {}".format(service))
+        if(os.path.exists("/usr/bin/juju")):
+            self.add_cmd_output([
+                "juju --version",
+                "juju -v status --format=tabular",
+                "juju -v status --format=json"
+            ])
+            for service in self.get_deployed_services():
+                self.add_cmd_output("juju get {}".format(service))
+            for service in self.get_deployed_services():
+                self.add_cmd_output("juju get-config {}".format(service))
+            for service in self.get_deployed_services():
+                self.add_cmd_output("juju get-constraints {}".format(service))
```

Comment:
Nack. This is running the exact same `juju status` command three times in a row and then parsing the result!

Just cache the returned value and use it in the subsequent loops:

```python
    services = self.get_deployed_services()
    for service in services:
        # do stuff
    for service in services:
       # do other stuff
```

Or better yet, since these loops are all iterating over the same thing, just get rid of them and have one:

```python
    for service in self.get_deployed_services():
        self.add_cmd_output("juju get {}".format(service))
        self.add_cmd_output("juju get-config {}".format(service))
        self.add_cmd_output("juju get-constraints {}".format(service))
```